File: TOH-Base/src/main/java/com/tohsoft/base/mvp/ui/BaseFragment.java
package com.tohsoft.base.mvp.ui;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.tohsoft.ads.AdsConfig;
import com.tohsoft.ads.AdsModule;
import com.tohsoft.base.mvp.utils.language.LocaleManager;
import com.utility.DebugLog;

/**
 * Created by Phong on 3/24/2017.
 */

public abstract class BaseFragment<P extends MvpPresenter> extends Fragment implements BaseMvpView {
    private BaseActivity mActivity;
    protected Context mContext;
    protected P mPresenter;

    protected abstract BasePresenter onRegisterPresenter();

    public BaseActivity getBaseActivity() {
        return mActivity;
    }

    private void initPresenter() {
        try {
            BasePresenter basePresenter = onRegisterPresenter();
            if (basePresenter != null) {
                basePresenter.attachView(this);
                mPresenter = (P) basePresenter;
            }
        } catch (Exception e) {
            DebugLog.loge(e);
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = getContext();
        setHasOptionsMenu(false);
        initPresenter();
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(LocaleManager.setLocale(context));
        if (context instanceof BaseActivity) {
            mActivity = (BaseActivity) context;
        }
    }

    protected void showBannerEmptyScreen(ViewGroup container) {
        AdsModule.getInstance().showBannerEmptyScreen(container);
    }

    protected void showPromotionView(View viewPromotionAds) {
        AdsModule.getInstance().showPromotionAdsView(viewPromotionAds);
    }

    protected void showPromotionAds() {
        AdsModule.getInstance().showPromotionAds();
    }

    @Override
    public void showLoading() {
        if (mActivity != null) {
            mActivity.showLoading();
        }
    }

    @Override
    public void showLoading(String message) {
        mActivity.showLoading(message);
    }

    @Override
    public void hideLoading() {
        if (mActivity != null) {
            mActivity.hideLoading();
        }
    }

    @Override
    public void showAlertDialog(String message) {
        if (mActivity != null) {
            mActivity.showAlertDialog(message);
        }
    }

    @Override
    public void hideAlertDialog() {
        if (mActivity != null) {
            mActivity.hideAlertDialog();
        }
    }

    @Override
    public void onDestroy() {
        if (mActivity != null) {
            mActivity.hideAlertDialog();
        }
        if (mActivity != null) {
            mActivity.hideLoading();
        }
        if (mPresenter != null) {
            mPresenter.detachView();
        }
        super.onDestroy();
    }

}

File: TOH-Base/src/main/java/com/tohsoft/base/mvp/ui/BaseFragment.kt
package com.tohsoft.base.mvp.ui

import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProviders
import com.tohsoft.ads.AdsModule
import com.tohsoft.base.mvp.utils.language.LocaleManager

/**
 * Created by Phong on 3/24/2017.
 */
abstract class BaseFragment : Fragment() {
    private var mBaseActivity: BaseActivity? = null
    protected lateinit var mContext: Context

    protected fun <M : ViewModel> getViewModel(modelClass: Class<M>): M {
        return ViewModelProviders.of(activity!!).get(modelClass)
    }

    protected fun <M : ViewModel> getActivityViewModel(modelClass: Class<M>): M {
        return ViewModelProviders.of(activity!!).get(modelClass)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mContext = context!!
        setHasOptionsMenu(false)
    }

    override fun onAttach(context: Context) {
        super.onAttach(LocaleManager.setLocale(context))
        if (context is BaseActivity) {
            mBaseActivity = context
        }
    }

    protected fun showBannerEmptyScreen(container: ViewGroup?) {
        AdsModule.getInstance().showBannerEmptyScreen(container)
    }

    protected fun showPromotionView(viewPromotionAds: View?) {
        AdsModule.getInstance().showPromotionAdsView(viewPromotionAds)
    }

    protected fun showPromotionAds() {
        AdsModule.getInstance().showPromotionAds()
    }

    fun showLoading() {
        mBaseActivity?.showLoading()
    }

    fun showLoading(message: String?) {
        mBaseActivity?.showLoading(message)
    }

    fun hideLoading() {
        mBaseActivity?.hideLoading()
    }

    fun showAlertDialog(message: String?) {
        mBaseActivity?.showAlertDialog(message)
    }

    fun hideAlertDialog() {
        mBaseActivity?.hideAlertDialog()
    }

    override fun onDestroy() {
        mBaseActivity?.hideAlertDialog()
        mBaseActivity?.hideLoading()
        super.onDestroy()
    }
}